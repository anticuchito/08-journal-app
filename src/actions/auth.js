import { firebase, googleAuthProvider } from "../firebase//firebase-config";
import Swal from 'sweetalert2';
import { types } from "../types/types";

import { startLoading, finishLoading } from "./ui";
//funciones que combierten las peticiones async en sync
export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());

    firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch((e) => {
        console.log(e);
        dispatch(finishLoading());
        Swal.fire('Error',e.message,'error');
      });
  };
};

export const startRegisterWhithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
      })
      .catch((e) => {
        console.log(e);
        Swal.fire('Error',e.message,'error')
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      }).catch(e => {
        dispatch(finishLoading());
        Swal.fire('Error',e.message,'error')
      })
  };
};
//funcion base para poder hacer el guardado con el reducer
export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});


export const  startLogout = () =>{
  return async(dispatch) =>{
    await firebase.auth().signOut();

    dispatch(logout());
  }
}

export const logout = () =>({
  type:types.logout
})