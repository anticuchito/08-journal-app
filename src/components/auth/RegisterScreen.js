import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import validator from "validator";
import { startRegisterWhithEmailPasswordName } from "../../actions/auth";
import { setError, removeError } from "../../actions/ui";
import { useForm } from "../../hooks/useForm";

/*
  {
    name'Hernando',
    email: 'nando@gmail.com,
    password:'123456'
    password:'123456'
  }

  use Form


  const HandleRegister = (e){
    console.log(name,email,password,password2)
  }
*/

export const RegisterScreen = () => {
  const dispatch = useDispatch();

  const { msgError } = useSelector((state) => state.ui);

  const [values, handleInputChange] = useForm({
    name: "Hernando",
    email: "nando@gmail.com",
    password: "123456",
    password2: "123456",
  });

  const { name, email, password, password2 } = values;

  const handleRegister = (e) => {
    e.preventDefault();
    if (isFormValid()) {
      dispatch(startRegisterWhithEmailPasswordName(email,password,name))
    }
  };

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError("name is required"));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setError("email is not valid"));
      return false;
    } else if (password !== password2 || password < 5) {
      dispatch(
        setError("password should at least 6 caracters and match each other")
      );
      return false;
    }
    dispatch(removeError());
    return true;
  };
  return (
    <div>
      <h3 className="auth__title">Register</h3>

      {   msgError &&
        (
          <div className="auth__alert-error">
            {msgError}
          </div>
        
        )
      }

      <form 
        onSubmit={handleRegister}
        className='animate__animated animate__fadeIn animate__faster'
        >
        <input
          type="text"
          placeholder="Name"
          name="name"
          className="auth__input"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />

        <input
          type="text"
          placeholder="Email"
          name="email"
          className="auth__input"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={password}
          onChange={handleInputChange}
        />

        <input
          type="password"
          placeholder="Confirm password"
          name="password2"
          className="auth__input"
          value={password2}
          onChange={handleInputChange}
        />

        <button type="submit" className="btn btn-primary btn-block mb-5">
          Register
        </button>

        <Link to="/auth/login" className="link mt-1">
          Already registered?
        </Link>
      </form>
    </div>
  );
};
