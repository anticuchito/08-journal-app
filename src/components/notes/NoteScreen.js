import React, { useEffect, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { activeNote, startDeleting } from "../../actions/notes";
import { useForm } from "../../hooks/useForm";
import { NotesAppBar } from "./NotesAppBar";

export const NoteScreen = () => {
  const { active: note } = useSelector((state) => state.notes);

  const [values, handleInputChange, reset] = useForm(note);

  const { body, title,id } = values;

  const activeId = useRef(note.id);

  const dispatch = useDispatch();
  useEffect(() => {
    if (activeId.current !== note.id) {
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activeNote(values.id, { ...values }));
  }, [values, dispatch]);

 const handleDelete = ()=>{
  dispatch(startDeleting(id));
 }

  return (
    <div className="notes__main-content">
      <NotesAppBar />

      <div className="notes__content">
        <input
          type="text"
          placeholder="Some awesome title"
          className="notes__title-input"
          autoComplete="off"
          value={title}
          onChange={handleInputChange}
          name="title"
        />

        <textarea
          placeholder="What happened today"
          className="notes__textarea"
          value={body}
          onChange={handleInputChange}
          name="body"
        ></textarea>

        {note.url && (
          <div className="notes__image">
            <img src={note.url} alt="imagen" />
          </div>
        )}
      </div>

      <button className="btn btn-danger" onClick={handleDelete}>
        Delete
      </button>
    </div>
  );
};
