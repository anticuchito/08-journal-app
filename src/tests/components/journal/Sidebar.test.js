import React from "react";
import { Provider } from "react-redux";

import configureStore from "redux-mock-store"; //ES6 modules
import thunk from "redux-thunk";
import "@testing-library/jest-dom";
import { SideBar } from "../../../components/journal/SideBar";
import { startLogout } from "../../../actions/auth";
import { startNewNote } from "../../../actions/notes";

const { mount } = require("enzyme");

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

jest.mock('../../../actions/auth',() => ({
   startLogout:jest.fn(),
}));

jest.mock('../../../actions/notes',() =>({
    startNewNote:jest.fn(),
    noteLogout:jest.fn()
}));


const initState = {
  auth: {
      uid:'123',
      name:'miguel'
 },
  ui: {
    loading: false,
    msgError: null,
  },
  notes:{
    notes:[{
        id:'123',
        body: 'test',
        title: 'test',
    }],
      active:null
  }};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
        <Provider store={store}>
                <SideBar />
        </Provider>
);


describe('pruebas en el <Sidebar />', () => {

    beforeEach(()=>{
        store = mockStore(initState);
        jest.clearAllMocks();
    })


    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de llamar logout', () => {
        wrapper.find('button').prop('onClick')();
        expect(startLogout).toHaveBeenCalled();
    });

    test('debe de llamar la accion del startNewnote', () => {
        wrapper.find('.journal__new-entry').prop('onClick')();
        expect(startNewNote).toHaveBeenCalled();
    })
    
    
    
})
