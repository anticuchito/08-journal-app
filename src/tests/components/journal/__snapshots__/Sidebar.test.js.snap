// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el <Sidebar /> debe de mostrarse correctamente 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
      />
      <span>
         
        miguel
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries "
  >
    <div
      className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
      onClick={[Function]}
    >
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        >
          test
        </p>
        <p
          className="journal__entry-content"
        >
          test
        </p>
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
          Thursday
        </span>
        <h4>
          3rd
        </h4>
      </div>
    </div>
  </div>
</aside>
`;
